### PR TITLE
`pd.eval`: `Series` names are now preserved even for `"numexpr"` engine.

### LINKED ISSUE #10239: BUG: numexpr eval doesn't hold column name via single column selection
```python
df = pd.DataFrame(np.random.randn(2, 3), columns=['A', 'B', 'C'])

numexpr.__version__
# '2.4'

# all resets column name

pd.eval('df.A')
# 0   -0.312782
# 1    1.204993
# dtype: float64

pd.eval('df["A"]')
# 0   -0.312782
# 1    1.204993
# dtype: float64

df.eval('A')
# 0   -0.312782
# 1    1.204993
# dtype: float64

# python engine looks OK
pd.eval('df["A"]', engine='python')
# 0   -0.312782
# 1    1.204993
# Name: A, dtype: float64

pd.eval('df.A', engine='python')
# 0   -0.312782
# 1    1.204993
# Name: A, dtype: float64

df.eval('A', engine='python')
# 0   -0.312782
# 1    1.204993
# Name: A, dtype: float64
```

### PR DESCRIPTION
**Pull request by domsmrz**
Series names are now preserved even when using
numexpr engine. Making the behavior consistent
with python engine.

- [x] closes #10239 (and its duplicate #58069)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v3.0.0.rst [hunk 1]
```diff
@@ -465,6 +465,7 @@ Styler
 Other
 ^^^^^
 - Bug in :class:`DataFrame` when passing a ``dict`` with a NA scalar and ``columns`` that would always return ``np.nan`` (:issue:`57205`)
+- Bug in :func:`eval` where the names of the ``Series`` were not preserved when using ``engine="numexpr"``. (:issue:`10239`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v3.0.0.rst:468]:
```suggestion
- Bug in :func:`eval` where the names of the :class:`Series` were not preserved when using ``engine="numexpr"``. (:issue:`10239`)
```

File: pandas/core/computation/align.py [hunk 2]
```diff
@@ -160,19 +160,24 @@ def align_terms(terms):
...
+        return np.result_type(terms.type), None, None
 
     # if all resolved variables are numeric scalars
     if all(term.is_scalar for term in terms):
-        return result_type_many(*(term.value for term in terms)).type, None
+        return result_type_many(*(term.value for term in terms)).type, None, None
+
+    # if all input series have a common name, propagate it to the returned series
+    names = {term.value.name for term in terms if isinstance(term.value, ABCSeries)}
+    name = next(iter(names)) if len(names) == 1 else None
```
Comment (mroeschke) [inline pandas/core/computation/align.py:173]:
```suggestion
    name = names.pop() if len(names) == 1 else None
```

File: pandas/core/computation/align.py [hunk 3]
```diff
@@ -200,7 +205,9 @@ def reconstruct_object(typ, obj, axes, dtype):
     res_t = np.result_type(obj.dtype, dtype)
 
     if not isinstance(typ, partial) and issubclass(typ, PandasObject):
-        return typ(obj, dtype=res_t, **axes)
+        if name is None:
+            return typ(obj, dtype=res_t, **axes)
+        return typ(obj, dtype=res_t, name=name, **axes)
```
Comment (mroeschke) [inline pandas/core/computation/align.py:210]: Shouldn't `name=None` also be OK? i.e. Can't this always be `return typ(obj, dtype=res_t, name=name, **axes)`?

File: pandas/tests/computation/test_eval.py [hunk 4]
```diff
@@ -737,6 +737,17 @@ def test_and_logic_string_match(self):
         assert pd.eval(f"{event.str.match('hello').a}")
         assert pd.eval(f"{event.str.match('hello').a and event.str.match('hello').a}")
 
+    def test_eval_keep_name(self, engine, parser):
+        df = Series([0.2, 1.5, 2.8], name="a").to_frame()
+        res = df.eval("a * a", engine=engine, parser=parser)
+        expected = df["a"] * df["a"]
```
[the remark this answers is not in the document]
Comment (domsmrz) [reply pandas/tests/computation/test_eval.py]: Done (also changed to `+`)

File: pandas/core/computation/align.py (hunk 3 above)
Comment (domsmrz) [reply pandas/core/computation/align.py:210]: Unfortunately no. In some cases `typ` is `DataFrame` (in particular cases like `pd.eval("df + df")`) in which case passing anything as `name=` (including `None`) would raise a `TypeError`

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @domsmrz

**Pull request merged**